File: 543.diameter-of-binary-tree/solution.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */
class Solution {
public:
    int diameterOfBinaryTree(TreeNode* root) {
        int i;
        return diameterOfBinaryTree(root, i);
    }
private:
    int diameterOfBinaryTree(TreeNode* root, int &height) {
        if (!root) {
            height = 0;
            return 0;
        }
        int heightleft, heightright;
        int leftdiameter = diameterOfBinaryTree(root->left, heightleft);
        int rightdiameter = diameterOfBinaryTree(root->right, heightright);
        height = max(heightleft, heightright) + 1;
        return max(heightleft + heightright, max(leftdiameter, rightdiameter));
    }
};

File: 543.diameter-of-binary-tree/solution.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Solution:
    def diameterOfBinaryTree(self, root: TreeNode) -> int:
        return self.diameterOfBinaryTree2(root)[0]

    def diameterOfBinaryTree2(self, root: TreeNode) -> (int, int):
        if not root:
            return 0, 0
        leftdiameter, heightleft = self.diameterOfBinaryTree2(root.left);
        rightdiameter, heightright = self.diameterOfBinaryTree2(root.right);
        return max(heightleft + heightright, leftdiameter, rightdiameter), max(heightleft, heightright) + 1;
